sycl/include/sycl/handler.hpp

```diff
@@ -384,6 +285,13 @@ template <int Dims> bool range_size_fits_in_size_t(const range<Dims> &r) {
   return true;
 }
 
+template <int Dims, typename LambdaArgType> struct TransformUserItemType {
+  using type = std::conditional_t<
+      std::is_convertible_v<nd_item<Dims>, LambdaArgType>, nd_item<Dims>,
+      std::conditional_t<std::is_convertible_v<item<Dims>, LambdaArgType>,
+                         item<Dims>, LambdaArgType>>;
+};
```

Comment:
https://github.com/intel/llvm/blob/259c433233e44ee9d4f0b91a3a10c28d854174a1/sycl/include/sycl/detail/type_traits.hpp#L385-L397 might be helpful.